### PR TITLE
HSEARCH-2612 Add JSR-352 Documentation Part 1

### PR DESCRIPTION
**Pull request by mincong-h**
This PR aims to prepare the documentation for JSR-352 integration, including:

- Move the JSR-352 user guide under folder `./documentation`.
- Update the obsolete documentation.
- Remove the wrong, misleading explanation about "Root Entity Types" in indexing, written in documentation.

However, it does not cover all the issues about JSR-352 documentation. Acutually, the following ones are not handled yet:

- Integrate the developer guide into the folder `./documentation`.
- Remove the wrong, misleading explanation about "Root Entity Types" in indexing, written in Javadoc. I'd like to do it after the merge of PR #1.

Related JIRA tickets:

- https://hibernate.atlassian.net/browse/HSEARCH-2612
- https://hibernate.atlassian.net/browse/HSEARCH-2626

### CODE REVIEW

**Comment by mincong-h** [comment]
Hi @yrodiere, please review my PR. I didn't have enough time to review the entire documentation this weekend, but I'll continue the next weekend. Have a nice week!

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 1]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+
+The integration with JSR-352, Batch Applications for the Java Platform, is in development. We do
+think we have the basics covered and we are looking for feedback.
+
+Hibernate Search provides a new mass indexing approach using batch job introduced by JSR352.
+It covers not only the existing functionality of the previous indexer, but also benefits the
+powerful standard based facilities of the Java Batch Platform (JSR-352), such as checkpoint and
+step partitioning. However, it requires a batch runtime for the execution:
+
+* For Java EE users, we use the batch runtime provided by your Java EE container.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: This line and the following should be rephrased; "we" don't provide the batch runtime, it's totally up to the user to choose one that fits their needs.

We could simply state that we provide integration to any CDI-enabled JSR-352 implementation (where the `EntityManagerFactory` is in the CDI context) out-of-the box, and that other implementations can be used with a bit more configuration on the user side.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 2]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+        .parameters()
+        .forEntity( MyEntity.class )
+        .build();
+long executionId = BatchRuntime
+        .getJobOperator()
+        .start( MassIndexingJob.NAME, jobProps );
+----
+====
+
+In the following paragraphs, we will explain why JSR-352 standard is a good choice, and how to
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Maybe I'm mistaken, but I think we generally avoid referring to ourselves in the documentation. So this sentence should be something like "The following paragraphs explain why the JSR-352 standard is a good choice, [...]".

Comment (yrodiere) [reply documentation/src/main/asciidoc/manual-index.asciidoc]: And, in fact, I think we could simply remove this parahtaph, it's not particularly useful.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 3]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+configure different parameters in order to bootstrap the mass-indexing job.
+
+==== Define your indexing scope
+Our job allows you to define your own indexing scope—you can start a full indexation or a partial
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Same here, avoid "our job". Maybe "the mass indexing job"?

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 4]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+indexation through 3 different methods: selecting the desired entity types, using HQL, or using
+Hibernate criteria.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: We should discuss the pros and cons of each method. In particular, no partitioning with HQL/Criteria (though this may change, see HSEARCH-2613).

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 5]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+// partial indexation using HQL
+Properties jobProps3 = MassIndexingJob
+        .parameters()
+        .forEntity( MyClass.class )
+        .restrictedBy( Restrictions.in( "name", "Foo", "Bar" ) )
+        .build();
+----
+====
+
+==== Parallelism
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Maybe "parallel indexing" rather than "parallelism? Same below.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 6]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+In order to maximize the performance, we highly recommend you to speed up the mass indexer using
+parallelism. Parallelism is activated by default. Under the standard JSR 352, the exact word is
+_"partitioning"_: The indexing step may run as multiple partitions, one per thread. Each partition
+has its own partition ID and parameters. If there’re more partitions than threads, partitions are\
+considered as a queue to consume: each thread can only run one partition at a time and won’t consume
+the next partition until the previous one is finished.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: This paragraph is a bit confusing. Maybe you shouldn't try to explain it in details, but only give a general idea of what's going on?

For instance:

 * "the indexing step" is confusing; users don't know there are multiple steps in our job, and they probably don't care.
 * "If there’re more partitions than threads, partitions are considered as a queue to consume" explains a corner case; it shouldn't be explained in an introduction (if at all).
 * "Under the standard JSR 352, the exact word is _"partitioning"_": it's not our business to teach JSR-352 to users (JSR-352 implementors probably do it better), and it makes the paragraph even more complex.

Maybe you could replace the paragraph with something like that:

```
For better performance, indexing is performed in parallel in multiple threads.
The set of entities to index is split into multiple partitions, each with a predefined number of elements,
and each thread processes one partition at a time.

The following sections detail how to tune the parallel execution.
```

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 7]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+
+===== Threads
+The maximum number of threads used by the job execution is defined through method `maxThreads()`.
+Within the N threads given, there’s 1 thread reserved for the core, so only N - 1 threads are
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Is this always true, regardless of the JSR-352 implementation?
Does it mean that setting maxThreads to 1 will not work?

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 8]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+available for different partitions. The default number of threads used in Hibernate Search is 10,
+which correspond to the max number of threads in JBeret. You can overwrite you maximum number used
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Please don't refer to JBeret in the docs, it's just one implementation among many others. Referring to it makes users of other JSR-352 implementations think that we just target JBeret, which isn't true.

Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: > but we cannot guarantee that your batch runtime can provide the requested number.

This is redundant with the note below. I would remove this and just keep the note.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 9]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+[NOTE]
+====
+Note that the batch runtime cannot guarantee the requested number of threads are available, it will
+use as it can up to the requested maximum. (JSR352 v1.0 Final Release, page 34)
+====
+
+This approach is optimized for best performance but requires to set the application in maintenance
+mode. Running a mass indexer in busy hours is not recommended. You also need to ensure the batch
+runtime has enough available threads for the execution, because all the batch jobs share the same
+thread pool.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: There are two topics here:

> This approach is optimized for best performance but requires to set the application in maintenance mode. Running a mass indexer in busy hours is not recommended.

This is general advice about how to run the job, and should in my opinion go to a separate section, preferably not in the middle of technical details.

>You also need to ensure the batch runtime has enough available threads for the execution, because all the batch jobs share the same thread pool.

This seems redundant with the note just above, or at least it's dealing with a very similar problem. Maybe we should add this to the note as a second paragraph?

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 10]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+
+===== Row per partitions
+Mass indexer defines dynamically the number of partitions, but you need to tell it the number of
+rows per partitions so that it can calculate the partitions by division. Each partition deals with
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: > Mass indexer defines dynamically the number of partitions [...] so that it can calculate the partitions by division

This feels like implementation details. Users don't care about how it's done internally, they just want to know how to use it.

What would you think of rephrasing the paragraph like this:

```
Each partition consists of a fixed number of elements to index.
You may tune exactly how many elements a partition will hold with `rowsPerPartition`.

[NOTE]
====
Each partition deals with one root entity type, so 2 different entity types will never run under the same partition.
====
```

Also, we should explain exactly how this number of rows per partition will influence the job. I mean, will it affect performance? Checkpoints? Something else? Why would setting this number to 10 instead of 10 million make a difference?

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 11]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .rowsPerPartition( 5000 )
+        ...
+----
+====
+
+==== Checkpoint
+Mass indexer supports checkpoint algorithm, which helps you to restart your suspended or failed job
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: `Mass indexer` => `The mass indexing job`, or something like that. We don't want users to confuse the job with the core `MassIndexer` class.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 12]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+from the last successful checkpoint. Assume that N is the value of checkpoint interval, then a
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: `Assume` => `Assuming`

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 13]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .checkpointInterval( 1000 )
+        ...
+----
+====
+
+Checkpoint a partition-scope algorithm, which means each partition has its own checkpoint counter
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: A word is missing here. I'd rephrase "Checkpoints are partition-scoped, meaning [...]"

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 14]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+and it is not shared with other partitions.
+
+
+==== Selecting the persistence unit (EntityManagerFactory)
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: I should restructure this section. Don't do it, I will, but I'm adding this comment as a reminder.

In particular:

 * the "Plain Java environment" section should go last
 * I should add a section about "other dependency injection mechanisms"
 * I should explain what I mean by "Plain Java environment" more precisely. In particular, it's not about "no CDI" but "no dependency injection mechanism".

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 15]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+long executionId = BatchRuntime
+        .getJobOperator()
+        .start( MassIndexingJob.NAME, jobProps );
+----
+====
+
+In the following paragraphs, we will explain why JSR-352 standard is a good choice, and how to
+configure different parameters in order to bootstrap the mass-indexing job.
+
+==== Define your indexing scope
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: Maybe use simpler wording here, such as "Define the entities to be indexed"? I'm worried about confusions between "indexing scope" and "entity manager factory reference scope", further down the page.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 16]
```diff
@@ -361,3 +361,226 @@ parameters through - see <<lucene-indexing-performance>> for more details.
 The MassIndexer uses a forward only scrollable result to iterate on the primary keys to be loaded,
 but MySQL's JDBC driver will load all values in memory; to avoid this "optimization" set
 `idFetchSize` to `Integer.MIN_VALUE`.
+
+
+[[jsr352-integration]]
+=== Integration with JSR-352
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc:367]: I'm not sure about the general layout of this documentation. These are the main problems I see:

 * We don't have a summary view of every parameter
 * Some parameters are not documented at all (`optimizeAfterPurge`, etc.)
 * Some concepts are explained across multiple sections, making it difficult to grasp them

Maybe the easier way out is to have two parts in the documentation:

 * a very simple "cheat sheet", i.e. a table with a "parameter name" column, a "builder method" column, and a "description" column (something along the lines of https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#table-worker-configuration). Of course, this table should be introduced properly ("Here is a table of... ", "The 'builder method' column contains...")
 * multiple sections providing in-depth documentation about a particular feature (potentially *multiple* parameters):
   * Defining the indexed entities
   * Retrieving the entity manager
   * Indexing in parallel (max threads + partitions + ...)
   * Dealing with stopped jobs (checkpoints)

**Review by yrodiere — CHANGES_REQUESTED** [review]
Here you go :)

About the developer documentation, be careful: our only documentation is a user guide. The developer documentation has nothing to do here. I'm not sure it's necessary to keep the whole developer documentation; maybe you could just take the parts relative to each particular component and move it to this component's Javadoc?

**Comment by mincong-h** [comment]
Thank you for your detailed review. It's really my pleasure to work with you. :+1:

> About the developer documentation, be careful: our only documentation is a user guide. The developer documentation has nothing to do here. I'm not sure it's necessary to keep the whole developer documentation; maybe you could just take the parts relative to each particular component and move it to this component's Javadoc?

Yes, you're right. I'll analyze the developer guide, and split them into two:

- add more detail into this user doc if there's anything useful
- add them into each particular component's Javadoc  if there's anything useful

I'll do it the next weekend in another PR.

As for the very simple "cheat sheet", i.e. a table with a "parameter name" column, a "builder method" column, and a "description" column... I didn't have enough time this weekend, will do it the next weekend, too.

File: documentation/src/main/asciidoc/manual-index.asciidoc (hunk 4 above)
Comment (mincong-h) [reply documentation/src/main/asciidoc/manual-index.asciidoc]: Thanks for your hint. Yes, the _HQL_ method does not support partitioning, but the _Criteria_ method does support partitioning. The key point is whether the selection, given by user, is orderable.

File: documentation/src/main/asciidoc/manual-index.asciidoc (hunk 7 above)
Comment (mincong-h) [reply documentation/src/main/asciidoc/manual-index.asciidoc]: > Is this always true, regardless of the JSR-352 implementation?

I didn't test it on other implementations except JBeret. According to spec 1.0, 11.7 Partitioned Chunk Processing (page 107), it seems that it is the case: there's one main thread `A` and x sub-threads `Px` for different partitions.

> Does it mean that setting maxThreads to 1 will not work?

It will work. In this case, we disable the multi-threading mechanism. This single thread will be used in both job operations and job executions. I updated the doc.

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 17]
```diff
@@ -361,3 +361,271 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+| Criteria
+| Partial Indexation
+| Supported
+| Supported
+|===
+
+==== Parallel indexing
+
+For better performance, indexation is performed in parallel using multiple threads. The set of
+entities to index is split into multiple partitions, each defined with its own partition ID and
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc]: > each defined with its own partition ID and parameters

I see what you did here :) But again, it's implementation details. If you feel mentioning "each with a predefined number of elements" isn't relevant/correct, maybe we should drop the "each defined..." part altogether?

File: documentation/src/main/asciidoc/manual-index.asciidoc [hunk 18]
```diff
@@ -361,3 +361,281 @@ parameters through - see <<lucene-indexing-performance>> for more details.
...
+====
+[source, JAVA]
+----
+MassIndexingJob.parameters()
+        .checkpointInterval( 1000 )
+        ...
+----
+====
+
+Checkpoint algorithms are partition-scoped, meaning each partition has its own checkpoint counter
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/manual-index.asciidoc:553]: What happens when the checkpoint interval is higher than the partition size? Do checkpoints become useless? Maybe we should trigger an exception in this case?

**Review by yrodiere — COMMENTED** [review]
Looks good, thanks :) I did a few adjustments and pushed them to your branch, feel free to comment or edit them.
Don't forget to pull before making new changes! Also, maybe you should rebase on my jsr352 branch (there has been some changes since you started your work).

I'll let you work on the rest before we merge.

**Comment by mincong-h** [comment]
Thanks Yoann. I'll take a look tomorrow or this Sunday!

File: documentation/src/main/asciidoc/manual-index.asciidoc (hunk 17 above)
Comment (mincong-h) [reply documentation/src/main/asciidoc/manual-index.asciidoc]: I agree with you, Yoann. It's an implementation detail and no need to mention here. Your proposition might lead to misunderstanding, because "element" sounds very similar to the word "item" of JSR-352. So I'll drop this sentence.

File: documentation/src/main/asciidoc/manual-index.asciidoc (hunk 18 above)
Comment (mincong-h) [reply documentation/src/main/asciidoc/manual-index.asciidoc:553]: Nice point. I'm not sure what happens in this case. Trigger an exception would be a good idea, issue created https://hibernate.atlassian.net/browse/HSEARCH-2638

**Comment by mincong-h** [comment]
Rebased. Ready to merge 😎

**Review by yrodiere — APPROVED** [review]

**Comment by yrodiere** [comment]
Merged, thanks! Don't forget to also submit a PR for the second part :) (I updated the descriptions of the two JIRA tickets)

**Comment by mincong-h** [comment]
Thanks Yoann. Yes, I'll submit the seconde part, and update the JIRA tickets myself the next time ;)
